README.md
<h1 align="center" style="">M1-gpufreq</h1>
<p align="center">
  Get the current frequency of your Apple M1 GPU.
</p>
<p align="center">
<a href="https://github.com/BitesPotatoBacks/M1-gpufreq/blob/main/LICENSE">
        <img alt="License" src="https://img.shields.io/github/license/BitesPotatoBacks/M1-gpufreq.svg"/>
    </a>
<!--     <a href="https://github.com/BitesPotatoBacks/M1-gpufreq/stargazers">
        <img alt="License" src="https://img.shields.io/github/stars/BitesPotatoBacks/M1-gpufreq.svg"/>
    </a> -->
    <a href="https://github.com/BitesPotatoBacks/M1-gpufreq/releases">
        <img alt="Releases" src="https://img.shields.io/github/v/release/BitesPotatoBacks/M1-gpufreq.svg"/>
    </a>
        <a href="https://cash.app/$bitespotatobacks">
        <img alt="License" src="https://img.shields.io/badge/donate-Cash_App-default.svg"/>
    </a>
    <!-- <a href="https://github.com/BitesPotatoBacks/osx-cpufreq/stargazers"><img alt="Stars" src="https://img.shields.io/github/stars/BitesPotatoBacks/osx-cpufreq.svg"/></a>-->
    <br>
</p>

## What It Does and How It Works
This project is designed to get the current frequency (or clock speed) of your Apple M1 GPU, without requiring `sudo` or a kernel extension. This near-impossible feat is achieved in a similar manner to how my [osx-cpufreq](https://github.com/BitesPotatoBacks/osx-cpufreq) project works, by accessing performance state information from `IOReport` and performing some calculations based on them during a specified time interval (default 1 second).

## Usage
### Preparation:
Download the precompiled binary from the [releases](https://github.com/BitesPotatoBacks/M1-gpufreq/releases), `cd` into your Downloads folder, and run these commands to fix the binary permissions:
```
chmod 755 ./M1-gpufreq
xattr -cr ./M1-gpufreq
```
Now you can simply run `./M1-gpufreq`.

### Example:
Here is an example running `./M1-gpufreq -l6 -i1` on an M1 Mac Mini during a Geekbench Compute run:
```
Name      Type      Max Freq     Active Freq    Freq %

GPU      Complex   1278.00 MHz     43.59 MHz     3.41%
GPU      Complex   1278.00 MHz   1153.91 MHz    90.29%
GPU      Complex   1278.00 MHz   1263.43 MHz    98.86%
GPU      Complex   1278.00 MHz    837.17 MHz    65.51%
GPU      Complex   1278.00 MHz     39.89 MHz     3.12%
GPU      Complex   1278.00 MHz   1235.26 MHz    96.66%
```

### Options
Available command line options are:
```
    -l <value> : loop output (0 = infinite)
    -i <value> : set sampling interval (may effect accuracy)
    -v         : print version number
    -h         : help
```

## Bugs and Issues
### Known Problems:
- Sampling interval is currently broken for some reason... run script with option `-i1` for now
- Support for M1 Pro/Max is unofficial

If any other bugs or issues are identified, please let me know!

## Support ❤️
If you would like to support me, you can donate to my [Cash App](https://cash.app/$bitespotatobacks).
<!-- 
## Changelog

```markdown
## [1.0.0] - Feb 4, 2022
- Initial Release
``` -->


CHANGELOG.md
## v1.0.1 (Feb 12, 2022)
#### Code Cleanup:
- Very minor improvements

#### Bug Fixes:
- Fixed possibility for script to ignore incorrect negative sampling intervals

## v1.0.0 (Feb 4, 2022)
- Initial Release


M1-gpufreq/main.m
//
//    MIT License
//
//    Copyright (c) 2022 BitesPotatoBacks
//
//    Permission is hereby granted, free of charge, to any person obtaining a copy
//    of this software and associated documentation files (the "Software"), to deal
//    in the Software without restriction, including without limitation the rights
//    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//    copies of the Software, and to permit persons to whom the Software is
//    furnished to do so, subject to the following conditions:
//
//    The above copyright notice and this permission notice shall be included in all
//    copies or substantial portions of the Software.
//
//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//    SOFTWARE.
//
//    Building:    clang main.m -fobjc-arc -arch arm64 -o M1-gpufreq -lIOReport -framework Foundation
//
//    Example Output:
//
//    Name      Type      Max Freq     Active Freq    Freq %
//
//    GPU      Complex   1278.00 MHz     43.59 MHz     3.41%
//    GPU      Complex   1278.00 MHz   1153.91 MHz    90.29%
//    GPU      Complex   1278.00 MHz   1263.43 MHz    98.86%
//    GPU      Complex   1278.00 MHz    837.17 MHz    65.51%
//    GPU      Complex   1278.00 MHz     39.89 MHz     3.12%
//    GPU      Complex   1278.00 MHz   1235.26 MHz    96.66%

#include <Foundation/Foundation.h>

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

#define VERSION         "1.1.0"
#define ERROR(ERR)      { printf("\e[1mM1-gpufreq\e[0m:\033[0;31m error:\033[0m %s\n", ERR); exit(-1); }

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

struct options
{
    float interval;
    int loop;
};

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

NSMutableArray * get_gpu_states(void);
NSMutableArray * get_gpu_nominal_freqs(void);

void  print_gpu_freq_info(int interval);
float return_gpu_active_freq(int interval);
float return_gpu_max_freq(void);

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

enum { kIOReportIterOk, kIOReportFormatState = 2 };

typedef struct IOReportSubscriptionRef* IOReportSubscriptionRef;
typedef CFDictionaryRef IOReportSampleRef;

extern IOReportSubscriptionRef IOReportCreateSubscription(void* a, CFMutableDictionaryRef desiredChannels, CFMutableDictionaryRef* subbedChannels, uint64_t channel_id, CFTypeRef b);
extern CFDictionaryRef IOReportCreateSamples(IOReportSubscriptionRef iorsub, CFMutableDictionaryRef subbedChannels, CFTypeRef a);

extern CFMutableDictionaryRef IOReportCopyChannelsInGroup(NSString*, NSString*, uint64_t, uint64_t, uint64_t);
extern CFMutableDictionaryRef IOReportCopyAllChannels(uint64_t, uint64_t);

typedef int (^ioreportiterateblock)(IOReportSampleRef ch);
extern void IOReportIterate(CFDictionaryRef samples, ioreportiterateblock);

extern int IOReportGetChannelCount(CFMutableDictionaryRef);
extern int IOReportChannelGetFormat(CFDictionaryRef samples);
extern long IOReportSimpleGetIntegerValue(CFDictionaryRef, int);

extern int IOReportStateGetCount(CFDictionaryRef);
extern uint64_t IOReportStateGetResidency(CFDictionaryRef, int);
extern NSString* IOReportStateGetNameForIndex(CFDictionaryRef, int);
extern NSString* IOReportChannelGetChannelName(CFDictionaryRef);
extern NSString* IOReportChannelGetGroup(CFDictionaryRef);
extern NSString* IOReportChannelGetSubGroup(CFDictionaryRef);
extern uint64_t IOReportArrayGetValueAtIndex(CFDictionaryRef, int);

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

int main(int argc, char * argv[])
{
    struct options  options;
    int             option;
    int             looper = 0;
    int             counterLooper = 0;
    
    options.loop        = NAN;
    options.interval    = 0.8;
    
    while((option = getopt(argc, argv, "i:l:vh")) != -1)
    {
        switch(option)
        {
            case 'l':   options.loop = atoi(optarg);
                        break;
                
            case 'i':   options.interval = atoi(optarg);
                        break;
                
            case 'v':   printf("\e[1mM1-gpufreq\e[0m:: version %s\n", VERSION);
                        return 0;
                        break;
                
            case 'h':   printf("Usage:\n");
                        printf("./M1-gpufreq [options]");
                        printf("    -l <value> : loop output (0 = infinite)\n");
                        printf("    -i <value> : set sampling interval (may effect accuracy)\n");
                        printf("    -v         : print version number\n");
                        printf("    -h         : help\n");
                        return 0;
                        break;
        }
    }
    
    if (options.interval < 0)
    {
        ERROR("sampling interval can not be a negative");
    }
    
    if (isnan(options.loop))
    {
        looper = 1;
    }
    else
    {
        if (options.loop == 0)
        {
            counterLooper = 1;
        }
        else
        {
            counterLooper = options.loop;
        }
    }
    
    printf("\e[1m\n%s%10s%14s%16s%10s\n\n\e[0m", "Name", "Type", "Max Freq", "Active Freq", "Freq %");
    
    while(looper < counterLooper)
    {
        print_gpu_freq_info(options.interval);
        
        if (options.loop != 0)
        {
            looper++;
        }
    }

    return 0;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

NSMutableArray * get_gpu_states(void)
{
    NSMutableArray * gpu_state_values = [NSMutableArray array];
    
    CFMutableDictionaryRef channels = IOReportCopyChannelsInGroup(@"GPU Stats", 0x0, 0x0, 0x0, 0x0);
    CFMutableDictionaryRef subbed_channels = NULL;
    
    IOReportSubscriptionRef subscription = IOReportCreateSubscription(NULL, channels, &subbed_channels, 0, 0);
    
    if (!subscription)
    {
        ERROR("error finding channel");
    }
    
    CFDictionaryRef samples = NULL;
    
    if ((samples = IOReportCreateSamples(subscription, subbed_channels, NULL)))
    {
        IOReportIterate(samples, ^(IOReportSampleRef ch)
        {
            NSString * channel_name = IOReportChannelGetChannelName(ch);
            NSString * subgroup = IOReportChannelGetSubGroup(ch);
            uint64_t value;

            for (int i = 0; i < IOReportStateGetCount(ch); i++)
            {
                if ([channel_name isEqual: @"GPUPH"] && [subgroup isEqual: @"GPU Performance States"])
                {
                    value = IOReportStateGetResidency(ch, i);
                    [gpu_state_values addObject:@(value)];
                }
            }
            
            return kIOReportIterOk;
        });
    }
    else
    {
        ERROR("error accessing performance state information");
    }
    
    if ([gpu_state_values count] == 0)
    {
        ERROR("missing performance state values");
    }
    
    return gpu_state_values;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

NSMutableArray * get_gpu_nominal_freqs(void)
{
    // need to figure out where to get these numbers so they don't have to be static in the code
    
    return [NSMutableArray arrayWithObjects: @"0", @"396", @"528", @"720", @"924", @"1128", @"1278", @"0", @"0", @"0", @"0", @"0", @"0", @"0", @"0", @"0", @"0", nil];
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

void print_gpu_freq_info(int interval)
{
    NSString * gpu_label = [NSString stringWithFormat:@"GPU %12s", "Complex"];
    
    float gpu_active_freq;
    float gpu_freq_percent;
    float gpu_max_freq = return_gpu_max_freq();
    
    gpu_active_freq = return_gpu_active_freq(interval);
    gpu_freq_percent = (gpu_active_freq / gpu_max_freq) * 100;

    if (gpu_freq_percent > 100)
    {
        gpu_freq_percent = 100;
        gpu_active_freq = gpu_max_freq;
    }
    
    if (gpu_freq_percent <= 0.009 || gpu_active_freq <= 0.009)
    {
        printf("%s%10.2f MHz%10.2f MHz %9s\n", [gpu_label UTF8String], gpu_max_freq, gpu_active_freq, "Idle");
    }
    else
    {
        printf("%s%10.2f MHz%10.2f MHz %8.2f%%\n", [gpu_label UTF8String], gpu_max_freq, gpu_active_freq, gpu_freq_percent);
    }
}

float return_gpu_active_freq(int interval)
{
    float first_sample_sum = 0;
    float last_sample_sum = 0;
    float sample_sum = 0;
    
    float state_percent = 0;
    float state_freq = 0;
    
    NSArray * gpu_nominal_freqs = get_gpu_nominal_freqs();
    
    NSMutableArray * first_sample = get_gpu_states();
    [NSThread sleepForTimeInterval:interval];
    NSMutableArray * last_sample = get_gpu_states();
    
    for (int i = 0; i < [first_sample count]; i++)
    {
        first_sample_sum += [first_sample[i] floatValue];
        last_sample_sum += [last_sample[i] floatValue];
    }
    
    sample_sum = last_sample_sum - first_sample_sum;
    
    for (int i = 0; i < [first_sample count]; i++)
    {
        state_percent = ([last_sample[i] floatValue] - [first_sample[i] floatValue]) / sample_sum;
        state_freq += state_percent * [gpu_nominal_freqs[i] floatValue];
    }
    
    if (isnan(state_freq))
    {
        return return_gpu_max_freq();
    }

    return state_freq;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

float return_gpu_max_freq(void)
{
//    NSArray * gpu_nominal_freqs = get_gpu_nominal_freqs();
//
//    return [gpu_nominal_freqs[[gpu_nominal_freqs count] - 1] floatValue];
    
    // doing this due poor handling of nominal freqs in get_gpu_nominal_freqs() causing issues
    return 1278;
}
